### ISSUE TITLE
"`NoPluginFoundForPrefixException`: No plugin found for prefix 'help' in the current project and in the plugin groups" when running on ACP

### ISSUE THREAD

**Issue by basil** [jenkins-infra/helpdesk#4021]
### Service(s)

Artifact-caching-proxy

### Summary

Compare https://ci.jenkins.io/job/Tools/job/bom/job/PR-3078/1/ (ACP) with https://ci.jenkins.io/job/Tools/job/bom/job/PR-3077/2/ (no ACP)—the former has failures in `pam-auth` and `sse-gateway` of this form:

```
2024-03-30 00:17:30.259+0000 [id=1]	INFO	o.j.t.t.m.ExternalMavenRunner#run: Running mvn -B -V -e -ntp -Dexpression=project.modules -Doutput=/home/jenkins/agent/workspace/Tools_bom_PR-3078/target/pct-work/sse-gateway-plugin/project.modules.log -q help:evaluate in /home/jenkins/agent/workspace/Tools_bom_PR-3078/target/pct-work/sse-gateway-plugin
Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae)
Maven home: /usr/share/apache-maven-3.9.6
Java version: 21.0.2, vendor: Eclipse Adoptium, runtime: /opt/jdk-21
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.10.209-198.858.amzn2.x86_64", arch: "amd64", family: "unix"
[ERROR] No plugin found for prefix 'help' in the current project and in the plugin groups [org.apache.maven.plugins, org.codehaus.mojo] available from the repositories [local (/home/jenkins/agent/workspace/Tools_bom_PR-3078@tmp/m2repo), aws-proxy (https://repo.aws.jenkins.io/)] -> [Help 1]
org.apache.maven.plugin.prefix.NoPluginFoundForPrefixException: No plugin found for prefix 'help' in the current project and in the plugin groups [org.apache.maven.plugins, org.codehaus.mojo] available from the repositories [local (/home/jenkins/agent/workspace/Tools_bom_PR-3078@tmp/m2repo), aws-proxy (https://repo.aws.jenkins.io/)]
	at org.apache.maven.plugin.prefix.internal.DefaultPluginPrefixResolver.resolve (DefaultPluginPrefixResolver.java:90)
	at org.apache.maven.lifecycle.internal.MojoDescriptorCreator.findPluginForPrefix (MojoDescriptorCreator.java:247)
	at org.apache.maven.lifecycle.internal.MojoDescriptorCreator.getMojoDescriptor (MojoDescriptorCreator.java:209)
	at org.apache.maven.lifecycle.internal.DefaultLifecycleTaskSegmentCalculator.calculateTaskSegments (DefaultLifecycleTaskSegmentCalculator.java:94)
	at org.apache.maven.lifecycle.internal.DefaultLifecycleTaskSegmentCalculator.calculateTaskSegments (DefaultLifecycleTaskSegmentCalculator.java:76)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:84)
	at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
	at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
	at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
	at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
	at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
	at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
	at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
	at java.lang.reflect.Method.invoke (Method.java:580)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:283)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:226)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:407)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:348)
 [ERROR] 
 [ERROR] Re-run Maven using the -X switch to enable full debug logging.
 [ERROR] 
 [ERROR] For more information about the errors and possible solutions, please read the following articles:
 [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/NoPluginFoundForPrefixException
 org.jenkins.tools.test.exception.PomExecutionException: mvn -B -V -e -ntp -Dexpression=project.modules -Doutput=/home/jenkins/agent/workspace/Tools_bom_PR-3078/target/pct-work/sse-gateway-plugin/project.modules.log -q help:evaluate in /home/jenkins/agent/workspace/Tools_bom_PR-3078/target/pct-work/sse-gateway-plugin failed with exit status 1
 	at org.jenkins.tools.test.maven.ExternalMavenRunner.run(ExternalMavenRunner.java:110)
 	at org.jenkins.tools.test.maven.ExpressionEvaluator.evaluateList(ExpressionEvaluator.java:58)
 	at org.jenkins.tools.test.PluginCompatTester.testPluginAgainst(PluginCompatTester.java:236)
 	at org.jenkins.tools.test.PluginCompatTester.testPlugins(PluginCompatTester.java:165)
 	at org.jenkins.tools.test.PluginCompatTesterCli.call(PluginCompatTesterCli.java:199)
 	at org.jenkins.tools.test.PluginCompatTesterCli.call(PluginCompatTesterCli.java:44)
 	at picocli.CommandLine.executeUserObject(CommandLine.java:2041)
 	at picocli.CommandLine.access$1500(CommandLine.java:148)
 	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
 	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
 	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
 	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
 	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
 	at picocli.CommandLine.execute(CommandLine.java:2170)
 	at org.jenkins.tools.test.CLI.main(CLI.java:13)
```

These consistent failures only occur with Artifact Caching Proxy (ACP). Furthermore, they did not occur on Wednesday, but do occur now starting on Friday, after ACP 1.0.0 was deployed.

The same failures also occurred on `mapdb-api`, `jjwt-api`, `pipeline-stage-step`, and `pipeline-milestone-step`. The one thing these plugins, `pam-auth`, and `sse-gateway` have in common is that the latest releases all have (or had) very ancient plugin parent POMs. I somehow worked around the problem in `mapdb-api`, `jjwt-api`, `pipeline-stage-step`, and `pipeline-milestone-step` by releasing refreshed versions of these plugins. I have no access to do this for `pam-auth` and `sse-gateway`.

In order for the next release of BOM to take place, either this bug needs to be fixed or the problem needs to be worked around by releasing refreshed versions of `pam-auth` and `sse-gateway`.

### Reproduction steps

Run two BOM builds with the `weekly-test` label, one with ACP and one without. The one with will show failures in `pam-auth` and `sse-gateway`; the one without will pass.

**Comment by dduportal** [comment]
Reproducted the issue locally and ran Maven with debug (`-X`) but without hiding transfert progress (removed the `-ntp` flag) and at first sight, Maven tries to retrieve its `help` plugin from the URL https://repo.do.jenkins.io/org/apache/maven/maven-help-plugin/ .

The plugin is located in https://repo1.maven.org/maven2/org/apache/maven/plugins/maven-help-plugin/: it means the ACP URL should be https://repo.do.jenkins.io/org/apache/maven/plugins/maven-help-plugin/.


I'm not sure why is the `/plugins` string removed by Maven, I guess it has something to do with the (settings.xml) custom profile currently setup for plugins. I'm focusing on this direction.

**Comment by dduportal** [comment]
OH: found it! 

We have https://repo.jenkins-ci.org/public/org/apache/maven/plugins/ which takes precedence in ACP over Central repository (see https://repo1.maven.org/maven2/org/apache/maven/plugins/)

The  `maven-metadata.xml` is from 2017 and only mentions 2 plugins, while Central repository clearly has way more elements.

It is most probably a remnant of when we used to mirror Maven Central repository in our artifactory.

While I'm investigatin in Artifactory where are these artifacts coming from (and search a solution to disable/remove them), Ping @daniel-beck does it ring a bell?

**Comment by dduportal** [comment]
> OH: found it!
> 
> We have https://repo.jenkins-ci.org/public/org/apache/maven/plugins/ which takes precedence in ACP over Central repository (see https://repo1.maven.org/maven2/org/apache/maven/plugins/)
> 
> The `maven-metadata.xml` is from 2017 and only mentions 2 plugins, while Central repository clearly has way more elements.
> 
> It is most probably a remnant of when we used to mirror Maven Central repository in our artifactory.
> 
> While I'm investigatin in Artifactory where are these artifacts coming from (and search a solution to disable/remove them), Ping @daniel-beck does it ring a bell?

Ping @jglick, as per repo.jenkins-ci.org, you were the person whom deployed a custom `maven-enforcer-plugin` (version `1.4.2-jenkins.1` which updated the metadata. Does it ring a bell?

(not there are other artifacts in this directory)


=> The goal would be to remove the `org/apache/maven/plugins` hierarchy from repo.jenkins-ci.org. WDYT y'all?

**Comment by basil** [comment]
> The goal would be to remove the org/apache/maven/plugins hierarchy

Given how many problems they have caused lately, I think there should be a systematic audit of all non-plugin (i.e., non HPI/JPI) artifacts in our `releases` repository to identify and quarantine unused artifacts. Snapshot artifacts and artifacts deployed 10 years ago are more likely to create problems than to solve them and should be archived into a separate repository.

**Comment by dduportal** [comment]
The tree `org/apache/maven` has been moved to `https://repo.jenkins-ci.org/artifactory/public-archives/org/apache/maven/` (with `public-archives` being a new local repository not integrated in `public`).

We now have a tree of empty directories left in `https://repo.jenkins-ci.org/public/org/apache/maven/plugins/` => I'm searching where are these directories and how to remove them

**Comment by dduportal** [comment]
The virtual repository `public` has been updated: the tree `org/apache/maven/public/**` has been added to the exclusion list.

https://repo.jenkins-ci.org/public/org/apache/maven/plugins/ => is now marked as HTTP/404 which is expected: ACP instances will now fall back to Maven central.

I have tested successfully a local build of the analysis-plugin. Currently checking on ci.jenkins.io with a bom build (and a plugin build).

@uhafner @MarkEWaite @basil can you confirm on your side?

**Comment by dduportal** [comment]
Triggered a bom build (`master` branch) and a Jenkins Core build (`branch`)  to check the results

**Comment by dduportal** [comment]
The 3 "test" builds I ran passed:

- jenkinsci/analysis-plugin (which was failing with ACP 1.0.0)
- bom and core master builds (but I did not see a failure for them).

I've opened https://github.com/jenkinsci/bom/pull/3089 to run the same kind of build as https://github.com/jenkinsci/bom/pull/3078

**Comment by MarkEWaite** [comment]
I've submitted a proposal to add the markdown-formatter plugin to the BOM managed set of plugins:

* https://github.com/jenkinsci/bom/pull/3090

That will run the weekly tests (about 165 plugins tested with Jenkins weekly).

**Comment by basil** [comment]
> The tree `org/apache/maven` has been moved to `https://repo.jenkins-ci.org/artifactory/public-archives/org/apache/maven/` (with `public-archives` being a new local repository not integrated in `public`).

But not https://github.com/jenkins-infra/helpdesk/issues/4021#issuecomment-2028807119?

> bom and core master builds (but I did not see a failure for them).

See https://github.com/jenkinsci/bom/issues/3080.

**Comment by dduportal** [comment]
> > The tree `org/apache/maven` has been moved to `https://repo.jenkins-ci.org/artifactory/public-archives/org/apache/maven/` (with `public-archives` being a new local repository not integrated in `public`).
> 
> But not [#4021 (comment)](https://github.com/jenkins-infra/helpdesk/issues/4021#issuecomment-2028807119)?

No we did not work on this: I believe it deserve a wider contribution group: the Jenkins infra team is missing a lot of knowledge and time to handle such a task by themselves. I believe it needs people with different profiles (including Jenkins Infra of course !) to drive this to completion.

=> this is why I tried a first (tiny) step to unblock the builds (with the awareness that, it might only pushes to the next dependency error). 

@basil WDYT about we start tracking the audit you mentioned as a separate task?

**Comment by dduportal** [comment]
> > bom and core master builds (but I did not see a failure for them).
> 
> See [jenkinsci/bom#3080](https://github.com/jenkinsci/bom/issues/3080).

Thanks for reminder, I forgot about the workaround and did not check in details. Let me update my draft PR to have the workaround removed to validate the changes I did.

**Comment by dduportal** [comment]
> > > bom and core master builds (but I did not see a failure for them).
> > 
> > 
> > See [jenkinsci/bom#3080](https://github.com/jenkinsci/bom/issues/3080).
> 
> Thanks for reminder, I forgot about the workaround and did not check in details. Let me update my draft PR to have the workaround removed to validate the changes I did.

The workaround has been reverted in https://github.com/jenkinsci/bom/pull/3089/commits/9a4020bdf32e91916bc2e49cebae3e7d08de1f9e and the PR checks are passing 🥳

**Comment by uhafner** [comment]
My plugin builds are working again! Thanks for fixing!

**Comment by jglick** [comment]
> Does it ring a bell?

No, sorry.
